File: YASAudio/Basic/YASAudioTime.m
//
//  YASAudioTime.m
//  Copyright (c) 2015 Yuki Yasoshima.
//

#import "YASAudioTime.h"
#import "YASMacros.h"
#import <mach/mach_time.h>

#pragma mark - C Function

static const mach_timebase_info_data_t *GetTimebaseInfo()
{
    static mach_timebase_info_data_t _timebaseInfo = {0, 0};
    if (_timebaseInfo.denom == 0) {
        mach_timebase_info(&_timebaseInfo);
    }
    return &_timebaseInfo;
}

#pragma mark - Main

@implementation YASAudioTime

- (instancetype)initWithAudioTimeStamp:(const AudioTimeStamp *)ts sampleRate:(Float64)sampleRate
{
    self = [super init];
    if (self) {
        _audioTimeStamp = *ts;
        _sampleRate = sampleRate;
    }
    return self;
}

- (instancetype)initWithHostTime:(UInt64)hostTime
{
    AudioTimeStamp ts = {
        .mHostTime = hostTime, .mFlags = kAudioTimeStampHostTimeValid,
    };

    return [self initWithAudioTimeStamp:&ts sampleRate:0];
}

- (instancetype)initWithSampleTime:(SInt64)sampleTime atRate:(Float64)sampleRate
{
    AudioTimeStamp ts = {
        .mSampleTime = sampleTime, .mFlags = kAudioTimeStampSampleTimeValid,
    };
    return [self initWithAudioTimeStamp:&ts sampleRate:sampleRate];
}

- (instancetype)initWithHostTime:(UInt64)hostTime sampleTime:(SInt64)sampleTime atRate:(Float64)sampleRate
{
    AudioTimeStamp ts = {
        .mHostTime = hostTime,
        .mSampleTime = sampleTime,
        .mFlags = kAudioTimeStampHostTimeValid | kAudioTimeStampSampleTimeValid,
    };
    return [self initWithAudioTimeStamp:&ts sampleRate:sampleRate];
}

+ (instancetype)timeWithAudioTimeStamp:(const AudioTimeStamp *)ts sampleRate:(Float64)sampleRate
{
    YASAudioTime *audioTime = [[self alloc] initWithAudioTimeStamp:ts sampleRate:sampleRate];
    return YASAutorelease(audioTime);
}

+ (instancetype)timeWithHostTime:(UInt64)hostTime
{
    YASAudioTime *audioTime = [[self alloc] initWithHostTime:hostTime];
    return YASAutorelease(audioTime);
}

+ (instancetype)timeWithSampleTime:(SInt64)sampleTime atRate:(Float64)sampleRate
{
    YASAudioTime *audioTime = [[self alloc] initWithSampleTime:sampleTime atRate:sampleRate];
    return YASAutorelease(audioTime);
}

+ (instancetype)timeWithHostTime:(UInt64)hostTime sampleTime:(SInt64)sampleTime atRate:(Float64)sampleRate
{
    YASAudioTime *audioTime = [[self alloc] initWithHostTime:hostTime sampleTime:sampleTime atRate:sampleRate];
    return YASAutorelease(audioTime);
}

+ (UInt64)hostTimeForSeconds:(NSTimeInterval)seconds
{
    UInt64 nanoSec = seconds * pow(10, 9);
    const mach_timebase_info_data_t *timebaseInfo = GetTimebaseInfo();
    return nanoSec * timebaseInfo->denom / timebaseInfo->numer;
}

+ (NSTimeInterval)secondsForHostTime:(UInt64)hostTime
{
    const mach_timebase_info_data_t *timebaseInfo = GetTimebaseInfo();
    Float64 nanoSec = hostTime * timebaseInfo->numer / timebaseInfo->denom;
    return nanoSec * pow(10, -9);
}

- (NSString *)description
{
    return
        [NSString stringWithFormat:
                      @"<%@: %p> hostTime=%@, isHostTimeValid=%@, sampleTime=%@, isSampleTimeValid=%@, sampleRate=%@",
                      NSStringFromClass(self.class), self, @(self.hostTime), @(self.isHostTimeValid),
                      @(self.sampleTime), @(self.isSampleTimeValid), @(self.sampleRate)];
}

- (YASAudioTime *)extrapolateTimeFromAnchor:(YASAudioTime *)anchorTime
{
    YASAudioTime *result = nil;

    if (anchorTime.isHostTimeValid && anchorTime.isSampleTimeValid && self.isSampleTimeValid) {
        NSTimeInterval anchorHostSeconds = [self.class secondsForHostTime:anchorTime.hostTime];
        NSTimeInterval sampleTimeSeconds =
            (_audioTimeStamp.mSampleTime / _sampleRate - anchorTime.audioTimeStamp.mSampleTime / anchorTime.sampleRate);
        UInt64 hostTime = [self.class hostTimeForSeconds:anchorHostSeconds + sampleTimeSeconds];
        result = [self.class timeWithHostTime:hostTime sampleTime:self.sampleTime atRate:self.sampleRate];
    }

    return result;
}

- (BOOL)isHostTimeValid
{
    return (_audioTimeStamp.mFlags & kAudioTimeStampHostTimeValid) != 0;
}

- (UInt64)hostTime
{
    return _audioTimeStamp.mHostTime;
}

- (BOOL)isSampleTimeValid
{
    return (_audioTimeStamp.mFlags & kAudioTimeStampSampleTimeValid) != 0;
}

- (SInt64)sampleTime
{
    return _audioTimeStamp.mSampleTime;
}

@end

File: Sources/audio/include/audio/common/yas_audio_time.cpp
//
//  yas_audio_time.mm
//

#include "yas_audio_time.h"

#include <cpp-utils/yas_cf_utils.h>
#include <mach/mach_time.h>

#include <exception>
#include <string>

using namespace yas;
using namespace yas::audio;

namespace yas::audio::time_utils {
static AudioTimeStamp make_time_stamp_from_host_time(uint64_t const host_time) {
    AudioTimeStamp timeStamp{0};
    timeStamp.mHostTime = host_time;
    timeStamp.mFlags = kAudioTimeStampHostTimeValid;
    return timeStamp;
}

static AudioTimeStamp make_time_stamp_from_sample_time(uint64_t const sample_time) {
    AudioTimeStamp timeStamp{0};
    timeStamp.mSampleTime = sample_time;
    timeStamp.mFlags = kAudioTimeStampSampleTimeValid;
    return timeStamp;
}

static AudioTimeStamp make_time_stamp(uint64_t const host_time, uint64_t const sample_time) {
    AudioTimeStamp timeStamp{0};
    timeStamp.mHostTime = host_time;
    timeStamp.mSampleTime = sample_time;
    timeStamp.mFlags = kAudioTimeStampSampleHostTimeValid;
    return timeStamp;
}

static bool is_equal(double const val1, double const val2, double const accuracy) {
    return ((val1 - accuracy) <= val2 && val2 <= (val1 + accuracy));
}

static bool is_equal(SMPTETime const &lhs, SMPTETime const &rhs) {
    return lhs.mSubframes == rhs.mSubframes && lhs.mSubframeDivisor == rhs.mSubframeDivisor &&
           lhs.mCounter == rhs.mCounter && lhs.mType == rhs.mType && lhs.mFlags == rhs.mFlags &&
           lhs.mHours == rhs.mHours && lhs.mMinutes == rhs.mMinutes && lhs.mSeconds == rhs.mSeconds &&
           lhs.mFrames == rhs.mFrames;
}

static bool is_equal(AudioTimeStamp const &lhs, AudioTimeStamp const &rhs) {
    return lhs.mHostTime == rhs.mHostTime && is_equal(lhs.mSampleTime, rhs.mSampleTime, 0.0001) &&
           lhs.mWordClockTime == rhs.mWordClockTime && is_equal(lhs.mRateScalar, rhs.mRateScalar, 0.0001) &&
           lhs.mFlags == rhs.mFlags && is_equal(lhs.mSMPTETime, rhs.mSMPTETime);
}

static mach_timebase_info_data_t const &timebase_info() {
    static mach_timebase_info_data_t _info{.numer = 0, .denom = 0};
    if (_info.denom == 0) {
        mach_timebase_info(&_info);
    }
    return _info;
}

static double const &host_time_to_seconds_rate() {
    static double _rate = 0.0;
    if (_rate == 0.0) {
        auto const &info = timebase_info();
        _rate = static_cast<double>(info.numer) / static_cast<double>(info.denom) / 1000000000.0;
    }
    return _rate;
}
}  // namespace yas::audio::time_utils

time::time(AudioTimeStamp const &time_stamp, double const sample_rate)
    : _time_stamp(time_stamp), _sample_rate(sample_rate) {
}

time::time(uint64_t const host_time) : _time_stamp(time_utils::make_time_stamp_from_host_time(host_time)) {
}

time::time(int64_t const sample_time, double const sample_rate)
    : _time_stamp(time_utils::make_time_stamp_from_sample_time(sample_time)), _sample_rate(sample_rate) {
}

time::time(uint64_t const host_time, int64_t const sample_time, double const sample_rate)
    : _time_stamp(time_utils::make_time_stamp(host_time, sample_time)), _sample_rate(sample_rate) {
}

bool time::is_host_time_valid() const {
    return this->_time_stamp.mFlags | kAudioTimeStampHostTimeValid;
}

uint64_t time::host_time() const {
    return this->_time_stamp.mHostTime;
}

bool time::is_sample_time_valid() const {
    return this->_time_stamp.mFlags | kAudioTimeStampSampleTimeValid;
}

int64_t time::sample_time() const {
    return this->_time_stamp.mSampleTime;
}

double time::sample_rate() const {
    return this->_sample_rate;
}

AudioTimeStamp time::audio_time_stamp() const {
    return this->_time_stamp;
}

bool time::operator==(time const &rhs) const {
    return time_utils::is_equal(this->_time_stamp, rhs._time_stamp) && this->_sample_rate == rhs._sample_rate;
}

bool time::operator!=(time const &rhs) const {
    return !(*this == rhs);
}

#pragma mark - global

uint64_t audio::host_time_for_seconds(double seconds) {
    return static_cast<uint64_t>(std::round(seconds / time_utils::host_time_to_seconds_rate()));
}

double audio::seconds_for_host_time(uint64_t host_time) {
    return static_cast<double>(host_time) * time_utils::host_time_to_seconds_rate();
}
